File: Linear_Data_Structures/Linked List/doubly_linked_list.py
"""
Doubly linked list implementation and demo
"""

class Node:
    """Node for doubly linked list"""
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None


class DoublyLinkedList:
    """Doubly linked list with basic operations."""
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def insert_at_head(self, data):
        new_node = Node(data)
        new_node.next = self.head
        if self.head:
            self.head.prev = new_node
        self.head = new_node
        if self.tail is None:
            self.tail = new_node
        self.size += 1

    def insert_at_tail(self, data):
        new_node = Node(data)
        if self.tail:
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = new_node
        else:
            self.head = self.tail = new_node
        self.size += 1

    def search(self, key):
        current = self.head
        while current:
            if current.data == key:
                return True
            current = current.next
        return False

    def delete(self, key):
        current = self.head
        while current:
            if current.data == key:
                if current.prev:
                    current.prev.next = current.next
                else:
                    self.head = current.next
                if current.next:
                    current.next.prev = current.prev
                else:
                    self.tail = current.prev
                self.size -= 1
                return True
            current = current.next
        return False

    def traverse_forward(self):
        current = self.head
        while current:
            yield current.data
            current = current.next

    def traverse_backward(self):
        current = self.tail
        while current:
            yield current.data
            current = current.prev

    def __len__(self):
        return self.size


if __name__ == "__main__":
    dl = DoublyLinkedList()
    for v in [1, 2, 3]:
        dl.insert_at_tail(v)
    print("Forward:", list(dl.traverse_forward()))
    print("Backward:", list(dl.traverse_backward()))

    dl.insert_at_head(0)
    print("After insert_at_head(0) Forward:", list(dl.traverse_forward()))

    dl.delete(2)
    print("After delete(2) Forward:", list(dl.traverse_forward()))
    print("Length:", len(dl))

File: Linear_Data_Structures/Linked List/doubly_linked_list.c
#include <stdio.h>
#include <stdlib.h>

/*
 * Doubly linked list implementation in C
 * Demo: insert head/tail, search, delete, traverse forward/backward
 */

typedef struct DNode {
    int data;
    struct DNode *prev;
    struct DNode *next;
} DNode;

DNode* create_node(int data) {
    DNode *n = (DNode*)malloc(sizeof(DNode));
    n->data = data;
    n->prev = n->next = NULL;
    return n;
}

void insert_at_head(DNode **head, DNode **tail, int data) {
    DNode *n = create_node(data);
    n->next = *head;
    if (*head) (*head)->prev = n;
    *head = n;
    if (*tail == NULL) *tail = n;
}

void insert_at_tail(DNode **head, DNode **tail, int data) {
    DNode *n = create_node(data);
    if (*tail) {
        (*tail)->next = n;
        n->prev = *tail;
        *tail = n;
    } else {
        *head = *tail = n;
    }
}

int search(DNode *head, int key) {
    DNode *cur = head;
    while (cur) {
        if (cur->data == key) return 1;
        cur = cur->next;
    }
    return 0;
}

int delete_value(DNode **head, DNode **tail, int key) {
    DNode *cur = *head;
    while (cur) {
        if (cur->data == key) {
            if (cur->prev) cur->prev->next = cur->next;
            else *head = cur->next;
            if (cur->next) cur->next->prev = cur->prev;
            else *tail = cur->prev;
            free(cur);
            return 1;
        }
        cur = cur->next;
    }
    return 0;
}

void traverse_forward(DNode *head) {
    DNode *cur = head;
    printf("[");
    while (cur) {
        printf("%d", cur->data);
        if (cur->next) printf(", ");
        cur = cur->next;
    }
    printf("]\n");
}

void traverse_backward(DNode *tail) {
    DNode *cur = tail;
    printf("[");
    while (cur) {
        printf("%d", cur->data);
        if (cur->prev) printf(", ");
        cur = cur->prev;
    }
    printf("]\n");
}

void free_list(DNode *head) {
    DNode *cur = head;
    while (cur) {
        DNode *next = cur->next;
        free(cur);
        cur = next;
    }
}

int main() {
    DNode *head = NULL, *tail = NULL;
    insert_at_tail(&head, &tail, 1);
    insert_at_tail(&head, &tail, 2);
    insert_at_tail(&head, &tail, 3);
    printf("Forward: ");
    traverse_forward(head);
    printf("Backward: ");
    traverse_backward(tail);

    insert_at_head(&head, &tail, 0);
    printf("After insert_at_head(0) Forward: ");
    traverse_forward(head);

    delete_value(&head, &tail, 2);
    printf("After delete 2 Forward: ");
    traverse_forward(head);

    free_list(head);
    return 0;
}
